File: app/src/main/java/com/laushkina/anastasia/psychosophy/domain/test/TestQuestion.java
package com.laushkina.anastasia.psychosophy.domain.test;

import android.os.Parcel;
import android.os.Parcelable;

public class TestQuestion implements Parcelable {
    private String text;
    private Function type;
    private TestAnswer answer;

    public TestQuestion(String text, Function type) {
        this.text = text;
        this.type = type;
    }

    public TestQuestion(String text, Function type, TestAnswer answer) {
        this.text = text;
        this.type = type;
        this.answer = answer;
    }

    protected TestQuestion(Parcel in) {
        text = in.readString();

        String function = in.readString();
        type = function == null ? null : Function.valueOf(function);

        String testAnswer = in.readString();
        answer = testAnswer == null ? null : TestAnswer.valueOf(testAnswer);
    }

    public static final Creator<TestQuestion> CREATOR = new Creator<TestQuestion>() {
        @Override
        public TestQuestion createFromParcel(Parcel in) {
            return new TestQuestion(in);
        }

        @Override
        public TestQuestion[] newArray(int size) {
            return new TestQuestion[size];
        }
    };

    public void setAnswer(TestAnswer answer) {
        this.answer = answer;
    }

    public String getText(){
        return text;
    }

    public Function getType(){
        return type;
    }

    public TestAnswer getAnswer(){
        return answer;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeString(text);
        parcel.writeString(type == null ? null : type.name());
        parcel.writeString(answer == null ? null : answer.name());
    }
}

File: app/src/main/java/com/laushkina/anastasia/psychosophy/domain/test/TestQuestion.kt
package com.laushkina.anastasia.psychosophy.domain.test

import android.os.Parcel
import android.os.Parcelable

class TestQuestion(val text: String,
                   val type: PsychoFunction?,
                   var answer: TestAnswer?) : Parcelable {

    constructor(text: String, type: PsychoFunction?) : this(text, type, null)

    constructor(parcel: Parcel) : this(
            parcel.readString() ?: "",
            functionFromString(parcel.readString()),
            answerFromString(parcel.readString())
    )

    companion object {
        @JvmField
        val CREATOR = object : Parcelable.Creator<TestQuestion> {
            override fun createFromParcel(parcel: Parcel): TestQuestion {
                return TestQuestion(parcel)
            }
            override fun newArray(size: Int) = arrayOfNulls<TestQuestion>(size)
        }

        private fun functionFromString(string: String?): PsychoFunction? {
            return if (string != null) PsychoFunction.valueOf(string) else null
        }

        private fun answerFromString(string: String?): TestAnswer? {
            return if (string != null) TestAnswer.valueOf(string) else null
        }
    }

    override fun writeToParcel(parcel: Parcel?, i: Int) {
        if (parcel == null) {
            return
        }
        parcel.writeString(text)
        parcel.writeString(type?.name)
        parcel.writeString(answer?.name)
    }

    override fun describeContents(): Int {
        return 0
    }



}